unittest/unit/update-bcd.ts

```diff
@@ -983,6 +1027,290 @@ describe('BCD updater', () => {
+      const report: Report = {
+        __version: '0.3.1',
+        results: {
+          'https://mdn-bcd-collector.appspot.com/tests/': [
+            {
+              name: 'api.AbortController',
+              exposure: 'Window',
+              result: false
+            }
+          ]
+        },
+        userAgent: firefox92UaString
+      };
+
+      const sm = getSupportMatrix([report], initialBcd.browsers, []);
+
+      const modified = update(finalBcd, sm, {});
+
+      assert.equal(modified, false, 'modified');
+      assert.deepEqual(finalBcd, initialBcd);
+    });
+
+    it('skips removed features', () => {
+      const initialBcd: any = {
+        api: {
+          AbortController: {
+            __compat: {
+              support: {
+                firefox: {version_removed: '91'}
+              }
+            }
+          }
+        },
+        browsers: {
+          firefox: {name: 'Firefox', releases: {91: {}, 92: {}, 93: {}}}
+        }
+      };
+      const finalBcd = clone(initialBcd);
+      const report: Report = {
+        __version: '0.3.1',
+        results: {
+          'https://mdn-bcd-collector.appspot.com/tests/': [
+            {
+              name: 'api.AbortController',
+              exposure: 'Window',
+              result: true
+            }
+          ]
+        },
+        userAgent: firefox92UaString
+      };
+
+      const sm = getSupportMatrix([report], initialBcd.browsers, []);
+
+      const modified = update(finalBcd, sm, {});
+
+      assert.equal(modified, false, 'modified');
+      assert.deepEqual(finalBcd, initialBcd);
+    });
+
+    it('persists non-default statements', () => {
+      const initialBcd: any = {
+        api: {
+          AbortController: {
+            __compat: {
+              support: {
+                firefox: {version_added: '91', prefix: 'moz'}
+              }
+            }
+          }
+        },
+        browsers: {
+          firefox: {name: 'Firefox', releases: {91: {}, 92: {}, 93: {}}}
+        }
+      };
+      const finalBcd = clone(initialBcd);
+      const report: Report = {
+        __version: '0.3.1',
+        results: {
+          'https://mdn-bcd-collector.appspot.com/tests/': [
+            {
+              name: 'api.AbortController',
+              exposure: 'Window',
+              result: true
+            }
+          ]
+        },
+        userAgent: firefox92UaString
+      };
+      const expectedBcd = clone(initialBcd);
+      expectedBcd.api.AbortController.__compat.support.firefox = [
+        {
+          version_added: '≤92'
+        },
+        {
+          prefix: 'moz',
+          version_added: '91'
+        }
+      ];
+
+      const sm = getSupportMatrix([report], initialBcd.browsers, []);
+
+      const modified = update(finalBcd, sm, {});
+
+      assert(modified, 'modified');
+      assert.deepEqual(finalBcd, expectedBcd);
+    });
+
+    it('overrides existing support information in response to negative test results', () => {
```

Comment:
It looks like that test would fail in the current implementation. This block of code appears to defer to the existing information in BCD:

https://github.com/foolip/mdn-bcd-collector/blob/19a623062ab0a82da8d3cc7c757452a5e408f877/update-bcd.ts#L400-L417

So I think that this would be best addressed in a separate patch.

Reply:
Wow this code is very hard to even parse. It's kind of weird that we use `versionMap` here when we've already used it to create `inferredStatements`. But yeah it looks like we're just going to bail if we have any non-null result from any browser version older than what we have in BCD. That happened in https://github.com/foolip/mdn-bcd-collector/pull/1904.

I agree sorting that out in this PR doesn't make sense.

Reply:
I've filed https://github.com/foolip/mdn-bcd-collector/issues/2348.